src/main/java/edu/kit/compiler/semantic/SymbolTable.java

```diff
@@ -0,0 +1,143 @@
+package edu.kit.compiler.semantic;
+
+import lombok.AllArgsConstructor;
+import lombok.Data;
+import lombok.NonNull;
+import lombok.RequiredArgsConstructor;
+
+import java.util.ArrayDeque;
+import java.util.LinkedHashMap;
+
+@Data
+public class SymbolTable  {
```

Comment:
As far as I can see, `currentScope` is not set during the construction of `SymbolTable`, right? So we have to call `enterScope` at least once before using the `SymbolTable` for insertion/lookup? If that's the case, we should probably add documentation about it to avoid null pointer exceptions.

Reply:
I actually find this behavior useful, but we should definitely document it.

Reply:
Yes the idea is that every variable has to be defined inside a scope. They get removed from the symbol table if you leave the scope that defined it. In code this is done by checking for `null` refernce to the current definition on `leaveScope()`